File: leetcode/77-combinations.py
# https://leetcode.com/problems/combinations/

from typing import List

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        combinations = list()
        combination = list()

        def backtracking(num):
            if len(combination) == k:
                combinations.append(combination.copy())
                return
            for num in range(num, n+1):
                combination.append(num)
                backtracking(num+1)
                combination.pop()
        
        backtracking(1)
        return combinations

File: leetcode/77-combinations.js
// https://leetcode.com/problems/combinations/

/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function(n, k) {
  const combinations = [];
  const combination = [];
  const backtracking = (num) => {
    if (combination.length === k) {
      combinations.push([...combination]);
      return;
    }

    if (num>4) {
      console.log(num);
      console.log(combination);
    }

    for (let i = num; i <= n; i++) {
      combination.push(i);
      backtracking(i+1);
      combination.pop();
    }
  };

  backtracking(1);
  return combinations;
};


console.log(combine(4, 2));